sys/net/gnrc/network_layer/ipv6/nib/nib.c

```diff
@@ -0,0 +1,630 @@
+#include <stdbool.h>
+
+#include "net/ipv6/addr.h"
+#include "net/gnrc/nettype.h"
+#include "net/gnrc/ipv6/netif.h"
+#include "net/gnrc/ipv6/nib.h"
+#include "net/gnrc/ndp2.h"
+#include "net/gnrc/pktqueue.h"
+#include "net/gnrc/sixlowpan/nd.h"
+#include "net/ndp.h"
+#include "net/sixlowpan/nd.h"
+
+#include "_nib-internal.h"
+#include "_nib-arsm.h"
+#include "_nib-6ln.h"
+#include "_nib-6lr.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+#if ENABLE_DEBUG
+#include "xtimer.h"
+#endif
+
+#if ENABLE_DEBUG
+static char addr_str[IPV6_ADDR_MAX_STR_LEN];
+#endif
+
+#if GNRC_IPV6_NIB_CONF_QUEUE_PKT
+static gnrc_pktqueue_t _queue_pool[GNRC_IPV6_NIB_NUMOF];
+#endif
+
+/**
+ * @internal
+ * @{
+ */
+static void _handle_nbr_sol(kernel_pid_t iface, const ipv6_hdr_t *ipv6,
+                            const ndp_nbr_sol_t *nbr_sol, size_t icmpv6_len);
+static void _handle_nbr_adv(kernel_pid_t iface, const ipv6_hdr_t *ipv6,
+                            const ndp_nbr_adv_t *nbr_adv, size_t icmpv6_len);
+
+static bool _resolve_addr(const ipv6_addr_t *dst, kernel_pid_t iface,
+                          gnrc_pktsnip_t *pkt, gnrc_ipv6_nib_nc_t *nce,
+                          _nib_onl_entry_t *entry);
+
+static void _handle_snd_na(gnrc_pktsnip_t *pkt);
+
+/* interface flag checks */
+#if GNRC_IPV6_NIB_CONF_ROUTER
+static inline bool _is_rtr(const gnrc_ipv6_netif_t *netif)
+{
+    return (netif->flags & GNRC_IPV6_NETIF_FLAGS_ROUTER);
+}
+#endif
+/** @} */
+
+void gnrc_ipv6_nib_init(void)
+{
+    evtimer_event_t *tmp;
+
+    mutex_lock(&_nib_mutex);
+    for (evtimer_event_t *ptr = _nib_evtimer.events;
+         (ptr != NULL) && (tmp = (ptr->next), 1);
+         ptr = tmp) {
+        evtimer_del((evtimer_t *)(&_nib_evtimer), ptr);
+    }
+    _nib_init();
+    mutex_unlock(&_nib_mutex);
+}
+
+void gnrc_ipv6_nib_init_iface(kernel_pid_t iface)
+{
+    _nib_iface_t *nib_iface;
+
+    assert(iface > KERNEL_PID_UNDEF);
+    DEBUG("nib: Initialize interface %u\n", (unsigned)iface);
+    mutex_lock(&_nib_mutex);
+    nib_iface = _nib_iface_get(iface);
+    assert(nib_iface != NULL);
+    /* TODO:
+     * - set link-local address here for stateless address auto-configuration
+     *   and 6LN
+     * - join solicited nodes group of link-local address here for address
+     *   resolution here
+     * - join all router group of link-local address here on router node here
+     * - become an router advertising interface here on non-6LR here */
+
+    _init_iface_arsm(nib_iface);
+    nib_iface->rs_sent = 0;
+    nib_iface->na_sent = 0;
+#if GNRC_IPV6_NIB_CONF_ROUTER
+    nib_iface->last_ra = UINT32_MAX;
+    nib_iface->ra_sent = 0;
+#endif
+    mutex_unlock(&_nib_mutex);
+}
+
+int gnrc_ipv6_nib_get_next_hop_l2addr(const ipv6_addr_t *dst,
+                                      kernel_pid_t iface, gnrc_pktsnip_t *pkt,
+                                      gnrc_ipv6_nib_nc_t *nce)
+{
+    int res = 0;
+
+    mutex_lock(&_nib_mutex);
+    do {    /* XXX: hidden goto ;-) */
```

Comment:
Since when do we use these *hidden gotos*? Frankly said, I do not like this idea. It 1) gives another level of indentation, 2) is an unconditional jump the same way as **goto** (so the same disadvantages apply) and 3) bloats the code by two more lines. I am sure you can re-arrange the code to not use this *hidden goto* and it still results in a more readable version (:

Reply:
> Since when do we use these hidden gotos? Frankly said, I do not like this idea.

Since @gebart proposed them as an alternative to "finally-gotos".

> 1) gives another level of indentation,

Since when is there this hatred for levels of indentations? We are not the Linux kernel (where this Coding Convention lead to very weird code btw).

> I am sure you can re-arrange the code to not use this hidden goto and it still results in a more readable version (:

I find it quite readable.

Reply:
```c
mutex_lock(&_nib_mutex);

if (!ipv6_addr_is_link_local(dst)) {
    /* TODO: Off-link next hop determination */
    mutex_unlock(&_nib_mutex);
    return -EHOSTUNREACH;
}

if ((iface == KERNEL_PID_UNDEF) ||
    !_resolve_addr(dst, iface, pkt, nce, _nib_onl_get(dst, iface))) {
    mutex_unlock(&_nib_mutex);
    return -EHOSTUNREACH;
}

return 0;
```

Isn't this so much simpler to read without any level of nestedness?!

Reply:
This adds 2 more function calls (`mutex_unlock()`), 6 in #7479 (`mutex_unlock()` + `gnrc_netif2_release()`), and even a more simpler version

```C
    int res = 0;

    mutex_lock(&_nib_mutex);
    if (!ipv6_addr_is_link_local(dst)) {    /* TODO: Prefix-based on-link determination */
        /* TODO: Off-link next hop determination */
        mutex_unlock(&_nib_mutex);
        return -EHOSTUNREACH;
    }
    if ((iface == KERNEL_PID_UNDEF) ||
        !_resolve_addr(dst, iface, pkt, nce,
                       _nib_onl_get(dst, iface))) {
        res = -EHOSTUNREACH;
    }
    mutex_unlock(&_nib_mutex);
    return res;
```

Adds 24 byte of ROM more on Cortex-M0.